File: libs/core-logic/src/lib/adapters/communication.adapter.ts
import { Observable } from 'rxjs';

/**
 * Абстрактный адаптер для связи с физическим устройством.
 * Определяет контракт, которому должны следовать платформенные реализации.
 * Ядро системы зависит ТОЛЬКО от этой абстракции.
 */
export abstract class CommunicationAdapter {
  /**
   * Состояние подключения к устройству.
   * true - подключено, false - не подключено.
   */
  abstract readonly connectionState$: Observable<boolean>;

  /**
   * Инициирует подключение к устройству по его имени или ID.
   * @param deviceIdentifier Имя или уникальный идентификатор устройства.
   * @returns Observable<boolean> - поток, который вернет true в случае успеха.
   */
  abstract connect(deviceIdentifier: string): Observable<boolean>;


  abstract disconnect(): void;
  
  abstract writeData(data: string): void;
}

File: lib/src/core_logic/adapters/communication_adapter.dart
import 'dart:async';
import 'package:flutter_blue_plus/flutter_blue_plus.dart';

/// Абстрактный класс CommunicationAdapter
/// Определяет контракт для взаимодействия с физическим устройством.
/// Ядро логики зависит только от этой абстракции.
abstract class CommunicationAdapter {
  /// Поток состояния подключения (true - подключено, false - нет).
  Stream<bool> get connectionState;
  Stream<List<ScanResult>> get scanResults;

  /// Метод для подключения к устройству по его имени.
  /// Возвращает Future<bool>, который завершается успешно (true), если подключение удалось.
  Future<bool> connect(BluetoothDevice device);
  Future<void> startScan({Duration timeout});
  Future<void> stopScan();


  Future<List<BluetoothService>> discoverServices();

  /// Метод для отключения от устройства.
  void disconnect();

  /// Метод для отправки данных (команд) на устройство.
  void writeData(BluetoothCharacteristic characteristic, String data);
  
  /// Метод для освобождения ресурсов.
  void dispose();
}